### PR TITLE
add fused dropout kernels

### PR DESCRIPTION
**Pull request by ngimel**
While waiting for dropout to be fully ported to ATen, here's performance fix for the most common dropout case. Dropout is still in python function, I just added efficient path to it. I could not make inplace work, because generator always generates `return self` for inplace function, and I need to return both original tensor and mask, so inplace goes on the existing pass. Even with non-inplace version, since mask is now a ByteTensor, memory used is just a little larger than for inplace dropout, due to savings on mask. 
Once dropout is moved to aten, these kernels still can be used for efficient implementation.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Also fusing ReLU + Dropout should be possible (https://discuss.pytorch.org/t/relu-dropout-inplace/13467)

**Comment by ngimel** [comment]
Build breaks because of #9435, I'll submit a fix for this PR, but #9435 also breaks other pytorch cpp extensions in the wild (e.g. apex)

**Comment by ngimel** [comment]
@vaidmkantorov that type of fusion will be better handled when random generation is available in the jit compiler.

**Comment by ngimel** [comment]
I don't know what's with rocm build failure, `Distributions.cu` seem to be using philox state in the same way, without special rocm handling.

File: torch/nn/_functions/dropout.py [hunk 1]
```diff
@@ -45,7 +55,13 @@ def forward(cls, ctx, input, p=0.5, train=False, inplace=False):
 
     @staticmethod
     def backward(ctx, grad_output):
-        if ctx.p > 0 and ctx.train:
+        if hasattr(ctx, 'use_fused_kernel') and ctx.use_fused_kernel:
```
Comment (apaszke) [inline torch/nn/_functions/dropout.py]: Why would it not have this attribute?

File: torch/nn/_functions/dropout.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import torch
 from torch.autograd.function import InplaceFunction
 from itertools import repeat
+from torch.autograd.function import once_differentiable
```
Comment (apaszke) [inline torch/nn/_functions/dropout.py]: Is this used?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -905,6 +909,11 @@
 - func: logsumexp_out(Tensor result, Tensor self, int64_t dim, bool keepdim=False) -> Tensor
   variants: function
 
+- func: masked_scale(Tensor self, Tensor mask, double scale) -> Tensor
```
Comment (apaszke) [inline aten/src/ATen/native/native_functions.yaml]: Hmm I'm not 100% sure if we want this in our API. Generally those cases should be served by `torch.where`. I know it's slightly less efficient, but in the end this should just be generated by the JIT. I guess we could have it as a private method (`_masked_scale`)

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -495,6 +495,10 @@
 - func: dot_out(Tensor result, Tensor self, Tensor tensor) -> Tensor
   variants: function
 
+- func: fused_dropout(Tensor self, double p, Generator* generator=nullptr) -> (Tensor, Tensor)
```
Comment (apaszke) [inline aten/src/ATen/native/native_functions.yaml]: This should probably be internal too.

File: aten/src/ATen/native/cuda/Dropout.cu [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+
+THCGenerator* THCRandom_getGenerator(THCState* state);
+
+namespace at{
+namespace native{
+
+namespace {
+
+//due to limitations of philox generator UNROLL has to be 4
```
Comment (apaszke) [inline aten/src/ATen/native/cuda/Dropout.cu]: Can you please comment what particular limitations do you have in mind in here?

File: aten/src/ATen/native/cuda/Dropout.cu [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+  accscalar_t pinv = accscalar_t(1)/p;
+  IndexType idx = blockIdx.x * blockDim.x + threadIdx.x;
+  curandStatePhilox4_32_10_t state;
+    curand_init(
+        seeds.first,
+        idx,
+        seeds.second,
+        &state);
+  IndexType rounded_size = ((totalElements - 1)/(blockDim.x*gridDim.x*UNROLL)+1)*blockDim.x*gridDim.x*UNROLL;
```
Comment (apaszke) [inline aten/src/ATen/native/cuda/Dropout.cu]: Can you please sprinkle some whitespace here and there? 😄 Or break this thing into multiple smaller variables/newlines

File: aten/src/ATen/native/cuda/Dropout.cu [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  });
+}
+} //anonymous namespace
+
+std::tuple<Tensor,Tensor>
+fused_dropout_cuda(const Tensor& self, double p, Generator * gen){
+  Tensor ret = at::empty_like(self);
+  Tensor mask = self.type().toScalarType(kByte).tensor(self.sizes());
+  const int64_t nelem = self.numel();
+  int64_t block_size = 256;
```
Comment (apaszke) [inline aten/src/ATen/native/cuda/Dropout.cu]: nit: `const`?

File: aten/src/ATen/native/cuda/Dropout.cu [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  unsigned int blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor/block_size;
+  dim3 dim_block(block_size);
+  dim3 grid((nelem + block_size -1)/block_size);
+  grid.x = std::min((unsigned int)at::cuda::getCurrentDeviceProperties()->multiProcessorCount * blocks_per_sm, grid.x);
+  int64_t nrep = ((nelem - 1)/(block_size*grid.x*UNROLL)+1)*UNROLL;
```
Comment (apaszke) [inline aten/src/ATen/native/cuda/Dropout.cu]: What's this? `nrep` is not super helpful

File: aten/src/ATen/native/cuda/Dropout.cu [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+      using accscalar_t = acc_type<scalar_t, true>;
+      accscalar_t pa = (accscalar_t)(p);
+      auto self_info = cuda::detail::getTensorInfo<scalar_t, unsigned int>(self);
+      auto ret_info = cuda::detail::getTensorInfo<scalar_t, unsigned int>(ret);
+      auto mask_info = cuda::detail::getTensorInfo<uint8_t, unsigned int>(mask);
+      self_info.collapseDims();
+      ret_info.collapseDims();
+      mask_info.collapseDims(); //ret and mask are collapsed to 1d contiguous tensor
+      switch (self_info.dims) {
+        case 1:
```
Comment (apaszke) [inline aten/src/ATen/native/cuda/Dropout.cu]: Why do we specialize on 1D strided tensors and not on the fully contiguous ones? Did that get removed at some point?

File: torch/nn/_functions/dropout.py (hunk 1 above)
Comment (ngimel) [reply torch/nn/_functions/dropout.py]: AlphaDropout and AlphaFeatureDropout override `forward` and thus don't set this attribute, but they share `backward` with `Dropout` class. So it's either set `use_fused_kernel` to `False` in AlphaDropout, or check for its existence in `backward`

File: torch/nn/_functions/dropout.py (hunk 2 above)
Comment (ngimel) [reply torch/nn/_functions/dropout.py]: No, I'll remove it

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (ngimel) [reply aten/src/ATen/native/native_functions.yaml]: Sure, will do `_masked_scale` and `_fused_dropout` - they'll still be available in python, right? I thought of using `where`, but it does not accept scalar argument as one of the tensors (it should, IMO, similar to comparison ops), and there's no way to scale in the same kernel, so it ends up being way slower. I totally agree this will be beside the point when randoms in jit are up and running, but until that, using more efficient dropout provides a few percent end-to-end speedup on some networks, so there's no reason not to use it.

File: aten/src/ATen/native/cuda/Dropout.cu (hunk 5 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/Dropout.cu]: Philox generated 128 random bits at a time, so to use them all, the unroll has to be 4. I'll put it in a comment.

File: aten/src/ATen/native/cuda/Dropout.cu (hunk 8 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/Dropout.cu]: Number of randoms generated by each thread, to increase philox counter in THC random state. Will put it in the comment.

File: aten/src/ATen/native/cuda/Dropout.cu (hunk 9 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/Dropout.cu]: Yes, fully contiguous specialization was removed from everywhere in one of @mruberry's PRs, the performance effects are negligible (one integer multiply vs no integer multiplies).

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (apaszke) [reply aten/src/ATen/native/native_functions.yaml]: They will, they just no longer be `torch.masked_scale`, but `torch._C._masked_scale`. I agree about allowing `where` to broadcast scalars.

File: torch/nn/_functions/dropout.py (hunk 1 above)
Comment (apaszke) [reply torch/nn/_functions/dropout.py]: I'd go for setting `use_used_kernel` to `False`. `hasattr` checks are quite slow.

**Comment by ezyang** [comment]
SGTM

**Comment by ailzhang** [comment]
cc: @ngimel for conflicting file

**Comment by ngimel** [comment]
@ailzhang done

**Comment by ngimel** [comment]
Anything I should do here to move forward?

**Comment by ezyang** [comment]
Nay, let's try to land it.

**Comment by ezyang** [comment]
There's an internal build system issue; the fix is easy though; I'll push it to this branch.
